VERSIONING.md

```diff
@@ -0,0 +1,220 @@
+              and will be applied to all existing stable modules as well as the newly
+              stable module being released.
+* Versioning of the associated [contrib
+  repository](https://github.com/open-telemetry/opentelemetry-collector-contrib) of
+  this project will be idiomatic of a Go project using [Go
+  modules](https://github.com/golang/go/wiki/Modules).
+    * [Semantic import
+      versioning](https://github.com/golang/go/wiki/Modules#semantic-import-versioning)
+      will be used.
+        * Versions will comply with [semver 2.0](https://semver.org/spec/v2.0.0.html).
+        * If a module is version `v2` or higher, the
+          major version of the module must be included as a `/vN` at the end of the
+          module paths used in `go.mod` files (e.g., `module
+          github.com/open-telemetry/opentelemetry-collector-contrib/processor/k8sprocessor/v2`, `require
+          github.com/open-telemetry/opentelemetry-collector-contrib/processor/k8sprocessor/v2 v2.0.1`) and in the
+          package import path (e.g., `import
+          "github.com/open-telemetry/opentelemetry-collector-contrib/processor/k8sprocessor/v2"`). This includes
+          the paths used in `go get` commands (e.g., `go get
+          github.com/open-telemetry/opentelemetry-collector-contrib/processor/k8sprocessor/v2@v2.0.1`.  Note there
+          is both a `/v2` and a `@v2.0.1` in that example. One way to think about
+          it is that the module name now includes the `/v2`, so include `/v2`
+          whenever you are using the module name).
+        * If a module is version `v0` or `v1`, do not include the major version
+          in either the module path or the import path.
+    * In addition to public APIs, telemetry produced by stable instrumentation
+      will remain stable and backwards compatible. This is to avoid breaking
+      alerts and dashboard.
+    * Modules will be used to encapsulate receivers, processor, exporters,
+      extensions, and any other independent sets of related components.
+        * Experimental modules still under active development will be versioned at
+          `v0` to imply the stability guarantee defined by
+          [semver](https://semver.org/spec/v2.0.0.html#spec-item-4).
+
+          > Major version zero (0.y.z) is for initial development. Anything MAY
+          > change at any time. The public API SHOULD NOT be considered stable.
+
+        * Mature modules for which we guarantee a stable public API and telemetry will
+          be versioned with a major version greater than `v0`.
+        * Experimental modules will start their versioning at `v0.0.0` and will
+          increment their minor version when backwards incompatible changes are
+          released and increment their patch version when backwards compatible
+          changes are released.
+        * Stable contrib modules cannot depend on experimental modules from this
+          project.
```

Comment:
Why is this rule not present for core? 

Also, I don't think this is necessarily a problem if the stable module does not use the unstable one in such a way that affects its public API. We use plenty of unstable dependencies on the core go.mod file, to give just one example: https://github.com/open-telemetry/opentelemetry-collector/blob/e8aeafa28e6557e93b41d587a0d8cdb94224bec7/go.mod#L34
and it works fine. Why is it worse to depend on experimental modules from within the project than to depend on experimental modules from outside the project?

Reply:
Depending on non-stable modules may cause dependency management problems. Let's assume an exporter depends on prom/common v0.25.0 and another one on prom/common v0.26.0. If prom/common does make a breaking change then the code will not be compilable.

I think we should evaluate every "experimental/unstable" dependency and ensure that we are always up-to-date if we decide to use one.

Reply:
I think that we can safely say that intra-project dependencies on unstable modules are prohibited.  It is within our sphere of influence entirely.  External dependencies that are not yet v1+ are more problematic as we may not have good alternatives and should be evaluated on a case-by-case basis.

Reply:
> I understand the motivation for the prohibition now, but I don't understand why it applies only to contrib. Could you clarify that @Aneurysm9?

Oh, I don't think it should.  If it is not clear that stable modules in the core repository cannot depend on unstable modules in the core repository I will seek to clarify that.

Reply:
Because we have decided that core should have a single module, I'm not sure it is necessary to state that stable core modules cannot depend on unstable core modules.  All core modules will be either unstable or stable.

I'm also not sure this statement is required for contrib modules anymore.  If there are not going to be unstable modules in the core repository once we declare any of them `v1`, then it is a moot point.  @bogdandrutu @mx-psi Do we think it is important to keep this statement until that point to provide guidance that no modules in contrib should have a `v1.x` version until the core module does?